Initial implementation (#1)
membraneframework-labs/membrane_style_transfer_plugin: README.md

# Membrane Style Transfer Plugin

[![Hex.pm](https://img.shields.io/hexpm/v/membrane_style_transfer_plugin.svg)](https://hex.pm/packages/membrane_style_transfer_plugin)
[![API Docs](https://img.shields.io/badge/api-docs-yellow.svg?style=flat)](https://hexdocs.pm/membrane_style_transfer_plugin)
[![CircleCI](https://circleci.com/gh/membraneframework/membrane_style_transfer_plugin.svg?style=svg)](https://circleci.com/gh/membraneframework/membrane_style_transfer_plugin)

This repository contains `Membrane.StyleTransfer` - `Membrane` filter performing style transfer on raw video frames.

It uses [Ortex](https://github.com/elixir-nx/ortex) to run models serialized in `.onnx` format and [Nx](https://github.com/elixir-nx/nx) to perform data pre- and postprocessing. 

It's a part of the [Membrane Framework](https://membrane.stream).

## Installation

The package can be installed by adding `membrane_style_transfer_plugin` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:membrane_style_transfer_plugin, github: "membraneframework-labs/membrane_style_transfer_plugin"}
  ]
end
```

## Usage

> Here we have 2 simple examples of usage `Membrane.StyleTransfer`. Both of them take video from the camera device available in your computer, perofrm style transfer and present the output video in the player.

Reviewer: ```suggestion
Here we have 2 simple examples of usage `Membrane.StyleTransfer`. Both of them take video from the camera device available in your computer, perform style transfer and present the output video in the player.
```

> On the beggining, install necessary dependencies.

Reviewer: ```suggestion
In the beginning, install necessary dependencies.
```

```elixir
Mix.install([
  {:membrane_style_transfer_plugin, github: "membraneframework-labs/membrane_style_transfer_plugin"},
  {:membrane_camera_capture_plugin, "~> 0.7.2"},
  {:membrane_ffmpeg_swscale_plugin, "~> 0.15.1"},
  {:membrane_sdl_plugin, "~> 0.18.2"}
])
```

> 

Reviewer: Some text comment could be helpful here, something like: "Then, run the following script":

```elixir
defmodule Example do
  use Membrane.Pipeline

  alias Membrane.FFmpeg.SWScale

  @impl true
  def handle_init(_ctx, opts) do
    height = opts[:image_height]
    width = opts[:image_width]

    spec =
      child(Membrane.CameraCapture)
      |> child(%SWScale.PixelFormatConverter{format: :I420})

>       |> child(%SWScale.Scaler{output_height: height, output_width: width})

Reviewer: I wonder why the `SWScale.Scaler` accepts only `:I420` on input, but it apparently does :D

      |> child(%SWScale.PixelFormatConverter{format: :RGB})
      |> child(%Membrane.StyleTransfer{style: opts[:style]})
      |> child(%SWScale.PixelFormatConverter{format: :I420})
      |> child(Membrane.SDL.Player)

    {[spec: spec], %{}}
  end
end
```

If you see that the latency of the output video is increasing, reduce `image_height` or/and `image_width`.

> If you see no increase in the latency, you can also increase the value passed in both options.

Reviewer: I am not sure if using `latency` here is the best option - how about referring to whether or not the frames are delivered in realtime?

> You can also change the style of played video by changing value passed in `:style` option. Available styles are: `:candy`, `:kaganawa`, `:mosaic`, `:mosaic_mobile`, `:picasso`, `:princess`, `:udnie` and `:vangogh`.

Reviewer: I would move those comments behind the: 
```
{:ok, _supervisor, pipeline} = Membrane.Pipeline.start_link(Example, [style: :vangogh, image_height: 400, image_width: 400])
```

```elixir
{:ok, _supervisor, pipeline} = Membrane.Pipeline.start_link(Example, [style: :vangogh, image_height: 400, image_width: 400])
```

> Example above used only one style at the time. Example below rotates styles atomatically every some time interval. 

Reviewer: ```suggestion
In the previous example, we applied just one style. However, in the following example, different styles are applied in rotation after fixed time intervals.
```

```elixir 
defmodule RotatingExample do
  use Membrane.Pipeline

  alias Membrane.FFmpeg.SWScale
  alias Membrane.StyleTransfer

  @impl true
  def handle_init(_ctx, opts) do
    height = opts[:image_height]
    width = opts[:image_width]
    first_style = :picasso

    spec =
      child(Membrane.CameraCapture)
      |> child(%SWScale.PixelFormatConverter{format: :I420})
      |> child(%SWScale.Scaler{output_height: height, output_width: width})
      |> child(%SWScale.PixelFormatConverter{format: :RGB})
      |> child(:style_tranfer, %StyleTransfer{style: first_style})
      |> child(%SWScale.PixelFormatConverter{format: :I420})
      |> child(Membrane.SDL.Player)

    {[spec: spec], %{current_style: first_style}}
  end

  @impl true
  def handle_playing(_ctx, state) do

>     interval = Membrane.Time.milliseconds(1_500)

Reviewer: Let's make this interval configurable as, for instance, a module attribute

It's a part of the [Membrane Framework](https://membrane.stream).

## Installation

The package can be installed by adding `membrane_style_transfer_plugin` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:membrane_style_transfer_plugin, github: "membraneframework-labs/membrane_style_transfer_plugin"}
  ]
end
```

## Usage

Here we have 2 simple examples of usage `Membrane.StyleTransfer`. Both of them take video from the camera device available in your computer, perform style transfer and present the output video in the player.

In the beginning, install necessary dependencies.

```elixir
Mix.install([
  {:membrane_style_transfer_plugin, github: "membraneframework-labs/membrane_style_transfer_plugin"},
  {:membrane_camera_capture_plugin, "~> 0.7.2"},
  {:membrane_ffmpeg_swscale_plugin, "~> 0.15.1"},
  {:membrane_sdl_plugin, "~> 0.18.2"}
])
```

Add implementation of the `Example` pipeline.

```elixir
defmodule Example do
  use Membrane.Pipeline

  alias Membrane.FFmpeg.SWScale

  @impl true
  def handle_init(_ctx, opts) do
    height = opts[:image_height]

>     width = opts[:image_width]

Reviewer: [NIT] How about renaming it into `output_height` and `output_width`? We are processing camera's video here, so I don't think that `image_*` fits.
It would also be more "self-explaining" if you used that name options in the "If you see that the latency of the output video is increasing, reduce `output_height` or/and `output_width`" sentence.
